python/rikai/spark/sql/callback_service.py

```diff
@@ -0,0 +1,82 @@
+#  Copyright 2021 Rikai Authors
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from py4j.java_gateway import CallbackServerParameters
+from pyspark.sql import SparkSession
+
+from rikai.logging import logger
+
+__all__ = ["init_cb_service"]
+
+
+def init_cb_service(spark: SparkSession):
+    jvm = spark.sparkContext._gateway
+    params = CallbackServerParameters(
+        daemonize=True,
+        daemonize_connections=True,
+        auth_token=jvm.gateway_parameters.auth_token,
+    )
+    jvm.start_callback_server(callback_server_parameters=params)
+    logger.info("Spark callback server started")
+
+    cb = CallbackService(spark)
+    cb.register()
+    logger.info("Rikai Python callback service registered")
+
+
+class CallbackService(object):
+    """:py:class:`CallbackService` allows SparkSessions' JVM to run
+    arbitrary code in SparkSession's python interpreter.
+
+    Notes
+    -----
+    Internal use only
+    """
+
+    def __init__(self, spark: SparkSession):
+        self.spark = spark
+
+    def __repr__(self):
+        return "PythonCbService"
+
+    def codegen(self, model, temporary: bool):
```

Comment:
where/when does this generated code get used?

Reply:
When a model first time introduced in the system, for example:

```
CREATE MODEL foo USING 'uri://to/path'
```

Or 

```
SELECT ML_PREDICT("uri://to/path", col) FROM t
```

In both cases, if the model needs to run in python,   `Model.pyClass.codegen()` will generate udfs, prepare enviornment and etc , and register the udfs into the main SparkSession in pyspark. So for the Spark SQL AST, it can replace the `ML_PREDICT` expression to a UDF that generated:

```
SELECT ML_PREDICT("uri://to/path/model", col) FROM t
=>
SELECT model_udf_tmp12345(col) FROM t
```

Each  `Model / Registry` will implement its own way to run the model, for example, we can implement a [vanilla pytorch inference](https://docs.databricks.com/_static/notebooks/deep-learning/pytorch-images.html) using `@pandas_udf` and `torch.DataLoader`,  or a UDF that sends http / gprc requests to an online inference server.